p2p/transport/webrtc/udpmux/mux.go

```diff
@@ -0,0 +1,247 @@
+package udpmux
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"strings"
+	"sync"
+
+	logging "github.com/ipfs/go-log/v2"
+	pool "github.com/libp2p/go-buffer-pool"
+	"github.com/pion/ice/v2"
+	"github.com/pion/stun"
+)
+
+var log = logging.Logger("mux")
+
+const ReceiveMTU = 1500
+
+var _ ice.UDPMux = &udpMux{}
+
+type ufragConnKey struct {
+	ufrag  string
+	isIPv6 bool
+}
+
+// udpMux multiplexes multiple ICE connections over a single net.PacketConn,
+// generally a UDP socket. The connections are indexed by (ufrag, IP address family)
+// and by remote address from which the connection has received valid STUN/RTC
+// packets. When a new packet is received on the underlying net.PacketConn, we
+// first check the address map to see if there is a connection associated with the
+// remote address. If found we forward the packet to the connection. If an associated
+// connection is not found, we check to see if the packet is a STUN packet. We then
+// fetch the ufrag of the remote from the STUN packet and use it to check if there
+// is a connection associated with the (ufrag, IP address family) pair. If found
+// we add the association to the address map. If not found, it is a previously
+// unseen IP address and the `unknownUfragCallback` callback is invoked.
+type udpMux struct {
+	mu                   sync.Mutex
```

Comment:
another very fragile and hard-to-reason use of a mutex here...

Reply:
The mux here is only used to protect two address maps.